flight/Modules/Logging/logging.c

```diff
@@ -232,107 +219,92 @@ static void loggingTask(void *parameters)
+				// log all UAVObjects, settings and data alike.
+				LL_FOREACH(log_info, info) {
+					// Iterate over all instances.
+					uint16_t numInstances = UAVObjGetNumInstances(info->obj);
+					for (int instId=0; instId<numInstances; instId++) {
+						UAVTalkSendObjectTimestamped(uavTalkCon, info->obj, instId, false, 0);
 					}
+					// use a random offset, so that not all objects are logged at the same time
+					info->next_log =  PIOS_Thread_Systime() + randomize_int(info->logging_period);
 				}
-
-				// Trigger logging for objects that are logged on change
-				flightstatus_updated = true;
-				waypoint_updated = true;
-
-				first_run = false;
-			}
-
-			// Log objects on change
-			if (flightstatus_updated){
-				UAVTalkSendObjectTimestamped(uavTalkCon, FlightStatusHandle(), 0, false, 0);
-				flightstatus_updated = false;
-			}
-
-			if (waypoint_updated && WaypointActiveHandle()){
-				UAVTalkSendObjectTimestamped(uavTalkCon, WaypointActiveHandle(), 0, false, 0);
-				waypoint_updated = false;
 			}
-
-			// Log very fast
-			UAVTalkSendObjectTimestamped(uavTalkCon, AccelsHandle(), 0, false, 0);
-			UAVTalkSendObjectTimestamped(uavTalkCon, GyrosHandle(), 0, false, 0);
-
-			// Log a bit slower
-			if ((i % 2) == 0) {
-				UAVTalkSendObjectTimestamped(uavTalkCon, AttitudeActualHandle(), 0, false, 0);
-				UAVTalkSendObjectTimestamped(uavTalkCon, MagnetometerHandle(), 0, false, 0);
-				UAVTalkSendObjectTimestamped(uavTalkCon, ManualControlCommandHandle(), 0, false, 0);
-				UAVTalkSendObjectTimestamped(uavTalkCon, ActuatorCommandHandle(), 0, false, 0);
-			}
-
-			// Log slower
-			if ((i % 10) == 1) {
-				UAVTalkSendObjectTimestamped(uavTalkCon, BaroAltitudeHandle(), 0, false, 0);
-				if (AirspeedActualHandle())
-					UAVTalkSendObjectTimestamped(uavTalkCon, AirspeedActualHandle(), 0, false, 0);
-				if (GPSPositionHandle())
-					UAVTalkSendObjectTimestamped(uavTalkCon, GPSPositionHandle(), 0, false, 0);
-				if (PositionActualHandle())
-					UAVTalkSendObjectTimestamped(uavTalkCon, PositionActualHandle(), 0, false, 0);
-				if (VelocityActualHandle())
-					UAVTalkSendObjectTimestamped(uavTalkCon, VelocityActualHandle(), 0, false, 0);
+			// Empty the queues
+			while(PIOS_Queue_Receive(shared_queue, &ev, 0))
+			LL_FOREACH(log_info, info) {
+				if (info->queue == shared_queue){
+					// The list is sorted based on the logging period, so we can stop as soon
+					// as we have an item without a private queue (as subsequent items won't 
+					// have a private queue either).
+					break;
+				}
+				while(PIOS_Queue_Receive(info->queue, &ev, 0)){};
 			}
+			LoggingStatsBytesLoggedSet(&written_bytes);
+			loggingData.Operation = LOGGINGSTATS_OPERATION_LOGGING;
+			LoggingStatsSet(&loggingData);
+			break;
+		case LOGGINGSTATS_OPERATION_LOGGING:
+			{
+				// Sleep between writing
+				PIOS_Thread_Sleep_Until(&now, LOGGING_PERIOD_MS);
+
+				// Log the objects with private queues
+				for (int i=0; i<PRIVATE_QUEUE_SIZE; i++){
```

Comment:
The fairness is currently implemented is that we iterate over the private queues several times and write one entry from each queue each time. So if the write speed to flash is not fast enough, the private queues will start dropping elements with the UAVOs that are being logged more frequently dropping more. E.g., if the flash is only fast enough to write 10 UAVOs of each registered type per second, UAVOs with lower logging intervals will be logged at approximately 10 UAVOs / second.

The long memory blanks that @kubark42 are actually time blanks, which occur when a flash page is erased (takes about 50ms). This is still happening unless the file system is freshly formatted (this is the reason the format option was added).

I agree with @mlyle that implementing the throttling in the UAVO callback layer would be the cleanest solution that could also be used elsewhere (e.g. in the telemetry module, which is currently very complex). The only concern is the increased memory consumption, i.e., I'm not sure if the F1 targets could handle it if it is implemented in the way that every callback has an interval associated with it. Maybe we could use a separate callback list for throttled callbacks to avoid this problem.

In general, a single queue would be cleaner. Especially if we implement throttled callbacks in the UAVO layer, as this would mean the logging module wouldn't need a linked list. This would also make it possible to determine the timestamp at the time an object is added to the queue, not at the time of writing to flash. If the queue is sufficiently deep, it will act as a buffer during periods when flash pages are erased and there will be no gaps in the timestamps.